docs/designs/ps_design.md

```diff
@@ -1,377 +1,146 @@
-
-Many `tf.keras.optimizer` subclasses, such as `Adam` and `Adagrad` allocate and manage additional variables associated with the variables to train.  These are called `Slots`.
-
-Embedding table slots are stored at KVStore, and other common parameter slots are stored and managed by `tf.keras.optimizer`.
-
-The embedding table slot is also a embedding table data structure. For example, a embedding table parameter with name `embedding_layer0`, we will create a corresponding `embedding_layer0-momentum` EmbeddingTable object in `KVStore.embedding_table_db`.
-
-We support async-SGD and sync-SGD both.
-
-There are two ways to support async-SGD:
- 
-- Calling `apply_gradient` inside `push_gradient` gRPC service.
-- Putting gradients into a gradient queue, and optimizer gets gradients from the queue immediately to `apply_gradient`.
-
-In the first way, there may be several gRPC threads running in parallel. This will introduce race condition on parameter updating, some gradients may be overwrited.
-
-The second way ensure each gradient could be applied, and decoupling these two procedures, `push_gradient` of worker and `apply_gradient` of optimizer. But the second way introduces more staleness in updating model, and may influence the final training accuracy.
-
-We may consider the second way later.
-
-In sync-SGD, optimizer needs to wait for a certain number of gradients, and then get the gradient after addition. We could implement a customized gradient queue structure to support such logic efficiently.
-
-The interface of gradient queue could be like this:
-
-```python
-class GradientQueue(object):
-    def __init__(self):
-        self.grad_queue = queue.Queue()
+    def __init__(self, layer_name, meta_info):
+        self.layer_name = layer_name
+        self.embeddings = {} # maps `id` to 1-D numpy.ndarray
+        self.meta_info = meta_info # meta_info is used for initializing embedding vectors
     
-    def get_gradient(self):
-        pass
+    def get(self, ids):
+        values = []
+        for id in ids:
+            if id not self.embeddings:
+                val = initialize_embedding_vector(self.meta_info)
+            else:
+                val = self.embeddings.get(id)
+            values.append(val)
+        return np.concatenate(values).reshape(len(ids), -1)
         
-    def put_gradient(self):
+    def set(self, ids, values):
         pass
 ```
 
-### RPC Service
-
-PS provides RPC service for workers.
+It is noticeable that we only have one interface for initializing parameters, `init_non_embedding_param`. This is because ElasticDL initializes embedding parameters lazily, i.e. initialize them when they are needed in the training process. When a worker sends a pull parameters request with some embedding ids to a PS instance, the PS instance calls the `get_embedding_param` function of its `Parameters` instance, and the `Parameters` instance will initialize embedding vectors for unknown ids, as shown in the above pseudocode. 
```

Comment:
We should tell why the embedding parameter is initialized lazily.

Reply:
Good point, added.